notifier/webhook/deliverer.go

```diff
@@ -0,0 +1,144 @@
+	kmgr *keymanager.Manager
+}
+
+// New returns a new webhook Deliverer
+func New(conf Config, client *http.Client, keymanager *keymanager.Manager) (*Deliverer, error) {
+	var c Config
+	var err error
+	if c, err = conf.Validate(); err != nil {
+		return nil, err
+	}
+	if client == nil {
+		client = http.DefaultClient
+	}
+	return &Deliverer{
+		conf: c,
+		c:    client,
+		kmgr: keymanager,
+	}, nil
+}
+
+func (d *Deliverer) Name() string {
+	return "webhook"
+}
+
+// sign will use the provided private key to sign and attach a jwt to the provided
+// request.
+//
+// a kid header specifying the private key which signed the jwt is attached as well.
+func (d *Deliverer) sign(ctx context.Context, req *http.Request, kp keymanager.KeyPair) error {
+	signer, err := jose.NewSigner(jose.SigningKey{Algorithm: jose.PS512, Key: kp.Private}, (&jose.SignerOptions{}).WithType("JWT"))
+	if err != nil {
+		return fmt.Errorf("failed to create jwt signer: %v", err)
+	}
+	now := jwt.NumericDate(time.Now().Unix())
+	expire := jwt.NumericDate(time.Now().Add(1 * time.Hour).Unix())
+	cl := jwt.Claims{
+		Subject:  "clair-notifier",
+		Issuer:   "clair-notifier",
+		Expiry:   &expire,
+		IssuedAt: &now,
+	}
+	kIDClaim := struct {
+		Kid string `json:"kid"`
+	}{
+		Kid: kp.ID.String(),
+	}
+	token, err := jwt.Signed(signer).Claims(cl).Claims(kIDClaim).FullSerialize()
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Authorization", "Bearer "+token)
+	req.Header.Set("kid", kp.ID.String())
+	return nil
+}
+
+// Deliver implements the notifier.Deliverer interface.
+//
+// Deliver POSTS a webhook data structure to the configured target.
+func (d *Deliverer) Deliver(ctx context.Context, nID uuid.UUID) error {
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "notifier/webhook/deliverer.Deliver").
+		Str("notification_id", nID.String()).
+		Logger()
+
+	callback := d.conf.callback
+	callback.Path = path.Join(callback.Path, nID.String())
+
+	wh := notifier.Callback{
+		NotificationID: nID,
+		Callback:       *callback,
+	}
+	b, err := json.Marshal(&wh)
+	if err != nil {
+		return err
+	}
+	buf := bytes.NewReader(b)
+
+	req := &http.Request{
+		URL:    d.conf.target,
+		Header: d.conf.Headers,
+		Body:   ioutil.NopCloser(buf),
+		Method: http.MethodPost,
+	}
+
+	// sign a jwt using key manager's private key
+	if d.conf.Signed {
+		kp, err := d.kmgr.KeyPair()
+		if err != nil {
+			return fmt.Errorf("configured for signing but no private key available: %v", err)
+		}
+		err = d.sign(ctx, req, kp)
+		if err != nil {
+			return fmt.Errorf("failed to sign request: %v", err)
+		}
+		log.Debug().Msg("successfully signed request")
+	}
+
+	log.Info().Str("notification_id", nID.String()).
+		Str("callback", callback.String()).
+		Str("target", d.conf.Target).
+		Msg("dispatching webhook")
+
+	resp, err := d.c.Do(req)
+	if resp != nil {
+		defer resp.Body.Close()
+	}
+	if err != nil {
+		return &clairerror.ErrDeliveryFailed{E: err}
+	}
+	if resp.StatusCode != http.StatusOK {
```

Comment:
The server will return OK and not Created or Accepted?

Reply:
@hdonnay I haven not worked with a webhook system where the server receiving a webhook does not response with 200OK.